bundles/core/org.eclipse.smarthome.core.thing.test/src/test/java/org/eclipse/smarthome/core/thing/internal/ThingManagerOSGiJavaTest.java

```diff
@@ -257,6 +265,228 @@ public void handleCommand(@NonNull ChannelUID channelUID, @NonNull Command comma
+        waitForAssert(() -> {
+            assertEquals(ThingStatus.UNINITIALIZED, thing.getStatus());
+        });
+
+        assertEquals(1, childHandlerInitializedSemaphore.availablePermits());
+
+        Thing thing2 = ThingBuilder.create(THING_TYPE_UID, THING_UID).withBridge(BRIDGE_UID).build();
+        managedThingProvider.update(thing2);
+        waitForAssert(() -> {
+            assertEquals(ThingStatus.ONLINE, thing2.getStatus());
+        });
+
+        // childHandlerInitialized(...) must be called
+        assertEquals(0, childHandlerInitializedSemaphore.availablePermits());
+
+        // thingUpdated(...) is not called
+        assertEquals(1, thingUpdatedSemapthore.availablePermits());
+    }
+
+    @Test
+    public void testChildHandlerInitialized_modifiedUninitializedThing() {
+        Semaphore childHandlerInitializedSemaphore = new Semaphore(1);
+        Semaphore thingUpdatedSemapthore = new Semaphore(1);
+
+        registerThingHandlerFactory(BRIDGE_TYPE_UID, bridge -> new BaseBridgeHandler((Bridge) bridge) {
+            @Override
+            public void handleCommand(@NonNull ChannelUID channelUID, @NonNull Command command) {
+            }
+
+            @Override
+            public void initialize() {
+                updateStatus(ThingStatus.ONLINE);
+            }
+
+            @Override
+            public void childHandlerInitialized(ThingHandler childHandler, Thing childThing) {
+                try {
+                    childHandlerInitializedSemaphore.acquire();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            };
+        });
+        registerThingHandlerFactory(THING_TYPE_UID, thing -> new BaseThingHandler(thing) {
+            @Override
+            public void handleCommand(@NonNull ChannelUID channelUID, @NonNull Command command) {
+            }
+
+            @Override
+            public void initialize() {
+                if (getBridge() == null) {
+                    throw new RuntimeException("Fail because of missing bridge");
+                }
+                updateStatus(ThingStatus.ONLINE);
+            }
+
+            @Override
+            public void thingUpdated(Thing thing) {
+                this.thing = thing;
+                try {
+                    thingUpdatedSemapthore.acquire();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            };
+        });
+
+        Bridge bridge = BridgeBuilder.create(BRIDGE_TYPE_UID, BRIDGE_UID).build();
+        managedThingProvider.add(bridge);
+        waitForAssert(() -> {
+            assertEquals(ThingStatus.ONLINE, bridge.getStatus());
+        });
+
+        Thing thing = ThingBuilder.create(THING_TYPE_UID, THING_UID).build();
+        managedThingProvider.add(thing);
+        waitForAssert(() -> {
+            assertEquals(ThingStatus.UNINITIALIZED, thing.getStatus());
```

Comment:
The bridge is not required, correct?
So, wouldn't it be possible that we miss the uninitialized status (timing)?
Shouldn't we wait for online as this status should be reached and the thing should stay in?

Reply:
It technically (currently) is not required. But if it's missing, the handler throws a RuntimeException, therefore the ThingManager puts it back into UNINITIALIZED/HANDLER_INITIALIZATION_ERROR. We can't miss it here.

Reply:
Ah ****, I missed the different `initialize` implementation as I tried to find the difference to the test below that waits for `ONLINE`.
Thanks and sorry :wink: